File: src/lib/sim/warmup.ts
// Welch's method + MSER-5 for warm-up cutoff detection.
// Browser-side mirror of sim-worker/sim_worker/warmup.py — used for explainer
// previews. Worker runs the authoritative computation.

export function welchMovingAverage(series: number[], window: number): number[] {
  const half = Math.floor(window / 2);
  const out: number[] = [];
  for (let i = half; i < series.length - half; i++) {
    let s = 0;
    for (let j = i - half; j <= i + half; j++) s += series[j];
    out.push(s / window);
  }
  return out;
}

/** MSER-5: find truncation that minimizes (sample variance / (n - d)^2). */
export function mser5(series: number[]): number {
  const n = series.length;
  if (n < 20) return 0;
  let bestD = 0;
  let bestVal = Infinity;
  // step by 5 for speed
  for (let d = 0; d < n - 10; d += 5) {
    const tail = series.slice(d);
    const m = tail.reduce((a, b) => a + b, 0) / tail.length;
    const v = tail.reduce((a, b) => a + (b - m) ** 2, 0) / tail.length;
    const score = v / ((n - d) * (n - d));
    if (score < bestVal) {
      bestVal = score;
      bestD = d;
    }
  }
  return bestD;
}

export function detectWarmup(series: number[]): number {
  if (series.length < 30) return 0;
  return mser5(series);
}

File: sim-worker/sim_worker/warmup.py
"""Warm-up detection — Welch's method + MSER-5 fallback."""

from __future__ import annotations

from typing import Sequence


def welch_moving_average(series: Sequence[float], window: int) -> list[float]:
    half = window // 2
    out: list[float] = []
    for i in range(half, len(series) - half):
        s = sum(series[i - half : i + half + 1])
        out.append(s / window)
    return out


def mser5(series: Sequence[float]) -> int:
    """Return truncation index d that minimizes var/(n-d)^2."""
    n = len(series)
    if n < 20:
        return 0
    best_d, best_val = 0, float("inf")
    for d in range(0, n - 10, 5):
        tail = series[d:]
        m = sum(tail) / len(tail)
        v = sum((x - m) ** 2 for x in tail) / len(tail)
        denom = (n - d) ** 2
        score = v / denom if denom else float("inf")
        if score < best_val:
            best_val = score
            best_d = d
    return best_d


def detect_warmup(series: Sequence[float], welch_window: int = 5) -> int:
    """Return the recommended warm-up cutoff (in steps)."""
    if len(series) < 30:
        return 0
    smooth = welch_moving_average(series, welch_window)
    return mser5(smooth)
